Add js unit tests

philipcaisip opened the pull request:
Unit tests for previously uncovered js code.

Found some small bugs in file/validate and date/isLeapYear. This fixes those too.

philcaisip committed:
Add unit tests for wc/file/validate

Increase code coverage by testing file input validation.
Partly based off of tests founnd in wc.file.size.test.mjs.
Leaves exception and prompt alert statements untested,
will likely need a mock to get that last bit of coverage.

philcaisip committed:
Fix file input accepted bug from previous commit

Remodel the accept function to optionally take the input files in its
own array as a second parameter, i.e. not bound to the given file input
element. This is to enable validation of files if they are given
separately from the element, which is sometimes the case with
validate.mjs use.

The separate file array parameter is checked first. If not given, then
the files already inside the element are checked.

Includes basic unit tests for the new functionality. Also now passes
the tests that were failing from the previous commit because of the bug.

philcaisip committed:
Add unit tests for isLeapYear.mjs

Add basic unit tests for leap year check, including the odd exception
years like 1900 and 3000.

Uncovered a small bug causing false-negatives due to unexpected type
behaviour from the divisible-by-100 check. This commit fixes that too.

philcaisip committed:
Merge branch 'hermione' into unit-test-phil

ricksbrown commented:
Hi @philipcaisip thanks for this.
See how the SonarCloud Analysis failed? Do you want to have a crack at fixing it?

Also, see how your PR includes some changes to files under `wcomponents-i18n`? What changed, and what made you decide to add those changes to this PR? Are you sure those changes belong here?

ricksbrown commented:
@philipcaisip before I take a look at this please get the SonarCloud Analysis pasing.

philcaisip committed:
Rollback i18n changes

philcaisip committed:
Merge branch 'hermione' into unit-test-phil

philcaisip committed:
Break up file array check in accepted.mjs

Refactors the file check in accepted.mjs as suggested in sonarcloud
analysis (to reduce cognitive complexity).

philcaisip committed:
Separate first file check in accepted.mjs

This is to break the file check up even more/reduce cognitive complexity
which is flagged as an issue in static code analysis.

philipcaisip commented:
Hi @ricksbrown, SonarCloud is passing now with no issues. Please take a look.

philcaisip committed:
Merge branch 'hermione' into unit-test-phil

ricksbrown commented on wcomponents-theme/src/main/js/wc/file/accepted.mjs:
```diff
@@ -23,45 +25,64 @@ import getMimeType from "wc/file/getMimeType.mjs";
+		return (!mimeType && !extension) ||
+			(extension && acceptedType.indexOf(extension) >= 0) ||
+			(mimeType && acceptedType.indexOf(mimeType) >= 0);
+	}
```
Missing semicolon

ricksbrown commented on wcomponents-theme/src/test/spec/wc.file.accepted.test.mjs:
```diff
@@ -84,4 +84,16 @@ describe("wc/file/accepted", function() {
+		expect(accepted(element, fileInfo)).toBeTrue();
+	});
+
+	it("does nto accept files given as parameter with mismatched types", function() {
```
if you are doing another commit may as well fix the typo - nto > not

ricksbrown commented on wcomponents-theme/src/test/spec/wc.file.validate.test.mjs:
```diff
@@ -0,0 +1,97 @@
+	}
+
+	beforeEach(() => {
+		testHolder = document.body.appendChild(document.createElement("div"));
```
It's no big deal, but since you are creating a new `testHolder` for each test, and leaving the previous testHolder lying around, maybe you should have an `afterEach` that removes them too, something like this:

```js
afterEach(() => {
		testHolder.parentElement.removeChild(testHolder);
```

ricksbrown commented on wcomponents-theme/src/main/js/wc/file/accepted.mjs:
```diff
@@ -23,45 +25,64 @@ import getMimeType from "wc/file/getMimeType.mjs";
  * @alias module:wc/file/accepted
  * @requires module:wc/file/getMimeType
  * @param {HTMLInputElement} element A "file input" element.
+ * @param fileInfo The array of files to check, inside an object under "files" property
```
update the JSDoc to reflect that this is an optional parameter `[fileInfo]`

ricksbrown commented on wcomponents-theme/src/main/js/wc/file/accepted.mjs:
```diff
@@ -23,45 +25,64 @@ import getMimeType from "wc/file/getMimeType.mjs";
+		return passed;
+	};
+
+	if (fileInfo) return getMimeType(fileInfo).every(checkFileArray);
```
While this works fine and there is nothing at all wrong with it, we inherited a style format that insists on curly braces even for single statements. The thinking behind such a rule is that it prevents future bugs, as someone comes along and adds another statement, forgetting to add the curly braces.

I'm not saying I agree with it, but that's how the rest of the codebase is written, so I suggest you add the curly braces to this one too.

ricksbrown commented on wcomponents-theme/src/main/js/wc/file/accepted.mjs:
```diff
@@ -23,45 +25,64 @@ import getMimeType from "wc/file/getMimeType.mjs";
+			extension = fileType.ext ? (`.${fileType.ext}`) : "",
+			passed = checkFileBasic(mimeType, extension);
+
+		if (passed) {
```
Do you think this code might be clearer without the `passed` variable? 
Like, we just `return true` instead from all the places, here, line 71 and line 78?
 For example:
```js
if (checkFileBasic(mimeType, extension)) { 
	return true;
}
```
```js
if (extension === next) {
	return true;
}
```

etc

philcaisip committed:
Clean up accepted.mjs code

Changes taken from suggestions in #1842.

ricksbrown reviewed (approved):
Nice work, thanks for this.
